### PR TITLE
[5.1] Do not allow empty cache paths

### LINKED ISSUE #14271: Generates compiled file on root directory when set compiled path doesn't exists
I cloned a lumen-based project and changed my storage path.
When I visit the homepage, it comes out with the error:
```
ErrorException in Filesystem.php line 81:
file_put_contents(/f536123a25b04be16752e137caae936b): failed to open stream: Permission denied
```
It's quite weird that cache file is going to store at root directory. I didn't set that kind of thing.
After debugging, I found out that the problem comes from [config/view.php#L31](https://github.com/laravel/laravel/blob/master/config/view.php#L31).
According to the PHP Document of [realpath](http://php.net/realpath) function, it will return false when the directory doesn't exists, which will be convert to empty string `""` when generating the compiled path at [View/Compilers/Compiler.php#44](https://github.com/laravel/framework/blob/5.2/src/Illuminate/View/Compilers/Compiler.php#L44).
I think the compiled path should be created when the application starts or it's better to dump more friendly message rather then compiling on the root directory.

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Replaces #14277. Closes #14271.

### CODE REVIEW

File: src/Illuminate/View/Compilers/Compiler.php [hunk 1]
```diff
@@ -26,9 +27,15 @@
      * @param  \Illuminate\Filesystem\Filesystem  $files
      * @param  string  $cachePath
      * @return void
+     *
+     * @throws \InvalidArgumentException
      */
     public function __construct(Filesystem $files, $cachePath)
     {
+        if (! $cachePath) {
+            throw new InvalidArgumentException('The cache path must be non-empty.');
```
Comment (fitztrev) [inline src/Illuminate/View/Compilers/Compiler.php]: "non-empty" might be ambiguous in this context.

Comment (GrahamCampbell) [inline src/Illuminate/View/Compilers/Compiler.php]: I think it's ok. `false` and `null` are "empty" in php.

Comment (fitztrev) [inline src/Illuminate/View/Compilers/Compiler.php]: I meant because it's referring to a path which might be empty (not contain any files).

Comment (vlakoff) [inline src/Illuminate/View/Compilers/Compiler.php]: Indeed, one of these may be better:

* `The cache path must be a non-empty value.`
* `The cache path must not be an empty value.`

Comment (fitztrev) [inline src/Illuminate/View/Compilers/Compiler.php]: It's been changed in d9ed009ee0e882933760514aaab9c8406dce11ad

**Pull request merged**